File: Server/php/service.php
<?php

class Service {
    const WILDCARD_CHARACTERS = "*";
    const ENABLE_PRIVILEGE = 1;
    const ENABLE_AUDIO_PRIVILEGE = 2;
    const ENABLE_VIDEO_PRIVILEGE = 4;

    private $channel_id;
    private $user_id;
    private $privilege;

    public function __construct(string $channel_id, string $user_id, $privilege = null) {
        $this->channel_id = $channel_id;
        $this->user_id = $user_id;
        $this->privilege = $privilege;
    }

    public static function create_service_only_with_user_id(string $user_id): self {
        return new self(self::WILDCARD_CHARACTERS, $user_id);
    }

    public static function create_service_only_with_channel_id(string $channel_id): self {
        return new self($channel_id, self::WILDCARD_CHARACTERS);
    }

    public function add_audio_publish_privilege() {
        if ($this->privilege === null) {
            $this->privilege = 0 | self::ENABLE_PRIVILEGE;
        }
        $this->privilege |= self::ENABLE_AUDIO_PRIVILEGE;
    }

    public function add_video_publish_privilege() {
        if ($this->privilege === null) {
            $this->privilege = 0 | self::ENABLE_PRIVILEGE;
        }
        $this->privilege |= self::ENABLE_VIDEO_PRIVILEGE;
    }

    public function pack(): string {
        $buf = '';

        $channel_id_bytes = $this->channel_id;
        $buf .= pack('N', strlen($channel_id_bytes));
        $buf .= $channel_id_bytes;
        
        $user_id_bytes = $this->user_id;
        $buf .= pack('N', strlen($user_id_bytes));
        $buf .= $user_id_bytes;
        
        if ($this->privilege !== null) {
            $buf .= pack('C', 1);
            $buf .= pack('N', $this->privilege);
        } else {
            $buf .= pack('C', 0);
        }
        return $buf;
    }

    public static function unpack($buffer): self {
        $channel_id_length = unpack('N', fread($buffer,4))[1];
        $channel_id = fread($buffer,$channel_id_length);

        $user_id_length = unpack('N', fread($buffer,4))[1];
        $user_id = fread($buffer,$user_id_length);


        $has_privilege = unpack('C', fread($buffer,1))[1];
        $privilege = null;
        if ($has_privilege === 1) {
            $privilege = unpack('N', fread($buffer,4))[1];
        }

        return new self($channel_id, $user_id, $privilege);
    }
}
?>

File: Server/python2/service.py
import io
import struct


class Service:
    WILDCARD_CHARACTERS = "*"
    ENABLE_PRIVILEGE = 1
    ENABLE_AUDIO_PRIVILEGE = 2
    ENABLE_VIDEO_PRIVILEGE = 4
    ENABLE_SCREEN_PRIVILEGE = 8

    def __init__(self, channel_id, user_id, privilege=None):
        self.channel_id = channel_id
        self.user_id = user_id
        self.privilege = privilege

    @classmethod
    def create_service_only_with_user_id(cls, user_id):
        return cls(cls.WILDCARD_CHARACTERS, user_id)

    @classmethod
    def create_service_only_with_channel_id(cls, channel_id):
        return cls(channel_id, cls.WILDCARD_CHARACTERS)

    def add_audio_publish_privilege(self):
        if self.privilege is None:
            self.privilege = 0 | self.ENABLE_PRIVILEGE

        self.privilege |= self.ENABLE_AUDIO_PRIVILEGE

    def add_video_publish_privilege(self):
        if self.privilege is None:
            self.privilege = 0 | self.ENABLE_PRIVILEGE

        self.privilege |= self.ENABLE_VIDEO_PRIVILEGE

    def add_screen_publish_privilege(self):
        if self.privilege is None:
            self.privilege = 0 | self.ENABLE_PRIVILEGE

        self.privilege |= self.ENABLE_SCREEN_PRIVILEGE

    def pack(self):
        buf = io.BytesIO()

        channel_id_bytes = self.channel_id.encode('utf-8')
        buf.write(struct.pack('>I', len(channel_id_bytes)))
        buf.write(channel_id_bytes)

        user_id_bytes = self.user_id.encode('utf-8')
        buf.write(struct.pack('>I', len(user_id_bytes)))
        buf.write(user_id_bytes)

        if self.privilege is not None:
            buf.write(struct.pack('>?I', True, self.privilege))
        else:
            buf.write(struct.pack('>?', False))
        return buf.getvalue()

    @staticmethod
    def unpack(buf):
        channel_id_length = struct.unpack('>I', buf.read(4))[0]
        channel_id = buf.read(channel_id_length).decode('utf-8')

        user_id_length = struct.unpack('>I', buf.read(4))[0]
        user_id = buf.read(user_id_length).decode('utf-8')

        has_privilege = struct.unpack('>?', buf.read(1))[0]
        privilege = struct.unpack('>I', buf.read(4))[0] if has_privilege else None
        return Service(channel_id, user_id, privilege)
